File: haskell/src/Solutions/Euler164.hs
module Solutions.Euler164 where
import Utils.Array(modifyArray)
import Control.Monad
import Control.Monad.ST
import Data.Array.ST

-- (length left, first two digits)
type C = (Int, Int, Int)

compute :: Int
compute = runST $ do
    arr <- newArray ((2, 0, 0), (20, 9, 9)) 0 :: ST s (STArray s C Int)
    forM_ [0..9] $ \i -> forM_ [0..9] $ \j -> when (i + j <= 9) (writeArray arr (2, i, j) 1)
    forM_ [3..20] $ \s -> forM_ [0..9] $ \i -> forM_ [0..9 - i] $ \j -> do
        plus <- readArray arr (s - 1, i, j)
        forM_ [0..9 - i - j] $ \d ->
            modifyArray arr (s, d, i) (+plus)
    sum <$> forM [1..9] (\i -> sum <$> forM [0..9] (\j -> readArray arr (20, i, j)))

euler164 :: IO String
euler164 = return $ show compute

File: rust/src/solutions/euler164.rs
pub fn euler164() -> String {
    // [length left - 2][first digit][second digit]
    let mut arr = vec![vec![vec![0u64; 10]; 10]; 19];
    for i in 0..10 {
        for j in 0..10 - i {
            arr[0][i][j] = 1;
        }
    }
    for s in 1..19 {
        for i in 0..10 {
            for j in 0..10 - i {
                let plus = arr[s - 1][i][j];
                for d in 0..10 - i - j {
                    arr[s][d][i] += plus;
                }
            }
        }
    }
    let mut ans = 0;
    for i in 1..10 {
        for j in 0..10 {
            ans += arr[18][i][j];
        }
    }
    ans.to_string()
}
